src/hotspot/share/utilities/moveBits.hpp

```diff
@@ -26,74 +27,156 @@
 #define SHARE_UTILITIES_MOVEBITS_HPP
 
 #include "metaprogramming/enableIf.hpp"
+#include "utilities/byteswap.hpp"
+#include "utilities/debug.hpp"
 #include "utilities/globalDefinitions.hpp"
 
+#include <cstdint>
 #include <type_traits>
 
+// T reverse_bits<T>(T)
+//
+// Reverses the bits in the integral value of type T.
+
 template <typename T>
-class ReverseBitsImpl {
-  static const size_t NB = sizeof(T) * BitsPerByte;
+using CanReverseBitsImpl = std::integral_constant<bool, (std::is_integral<T>::value &&
+                                                         (sizeof(T) == 1 ||
+                                                          sizeof(T) == 2 ||
+                                                          sizeof(T) == 4 ||
+                                                          sizeof(T) == 8))>;
+
+template <typename T, size_t N = sizeof(T)>
+struct ReverseBitsImpl;
+
+template <typename T, ENABLE_IF(CanReverseBitsImpl<T>::value)>
+ALWAYSINLINE T reverse_bits(T x) {
+  return ReverseBitsImpl<T>{}(x);
+}
+
+/*****************************************************************************
+ * Fallback
+ *****************************************************************************/
 
-  static_assert((NB == 8) || (NB == 16) || (NB == 32) || (NB == 64),
-                "unsupported size");
+template <typename T>
+struct ReverseBitsFallbackImpl {
+ private:
+  static constexpr size_t NB = sizeof(T) * BitsPerByte;
 
   // The unsigned integral type for calculations.
   using I = std::conditional_t<NB <= 32, uint32_t, uint64_t>;
 
-  static const I rep_5555 = static_cast<I>(UCONST64(0x5555555555555555));
-  static const I rep_3333 = static_cast<I>(UCONST64(0x3333333333333333));
-  static const I rep_0F0F = static_cast<I>(UCONST64(0x0F0F0F0F0F0F0F0F));
-  static const I rep_00FF = static_cast<I>(UCONST64(0x00FF00FF00FF00FF));
-  static const I rep_FFFF = static_cast<I>(UCONST64(0x0000FFFF0000FFFF));
+  static constexpr I rep_5555 = static_cast<I>(UINT64_C(0x5555555555555555));
+  static constexpr I rep_3333 = static_cast<I>(UINT64_C(0x3333333333333333));
+  static constexpr I rep_0F0F = static_cast<I>(UINT64_C(0x0F0F0F0F0F0F0F0F));
 
-public:
+ public:
+  STATIC_ASSERT(CanReverseBitsImpl<T>::value);
 
-  static constexpr T reverse_bits_in_bytes(T v) {
+  ALWAYSINLINE T operator()(T v) const {
     // Based on Hacker's Delight Section 7-1
-    auto x = static_cast<I>(v);
+    I x = static_cast<I>(v);
     x = ((x & rep_5555) << 1) | ((x >> 1) & rep_5555);
     x = ((x & rep_3333) << 2) | ((x >> 2) & rep_3333);
     x = ((x & rep_0F0F) << 4) | ((x >> 4) & rep_0F0F);
-    return static_cast<T>(x);
+    return byteswap<T>(static_cast<T>(x));
```

Comment:
If we happen to use this on a platform that has hardware support for bitswap, and the implementation
of byteswap is an intrinsic, it seems like there might be a risk of interfering with the pattern recognition
that could transform into the hardware-supported method.

Reply:
Looks like neither GCC nor Clang are good at recognizing bit reverse, even with the existing implementation. At least based on compiler explorer output. The only way to get Clang to generate `RBIT` on AArch64 is to use `__builtin_bitreverse` or `__rbit` from `<arm_acle.h>`. GCC will also only generate `RBIT` when using `__rbit` from `<arm_acle.h>`. Using `__rbit` won't work to well because we would have to use preprocessor detection to detect whether the target actually has the instruction first. There does not seem like a consistent way to do this across GCC/Clang.